### PR TITLE
Overloaded WebClient's methods to accept query parameters

### LINKED ISSUE #3915: Provide a way to send query strings using `QueryParams` with `WebClient`
When sending query strings in a path using `WebClient`, users have to manually concatenate query params. For example:
```java
WebClient client = ...;
QueryParams params = QueryParams.of("foo", "bar");
client.get("/foo?" + params.toQueryString());
```
It would be convenient to automatically append the query parameters to the path.
```java
client.get("/foo", params);
client.post("/bar", params, content);
```

### PR DESCRIPTION
**Pull request by nvidyala**
Motivation:

WebClient's HTTP-specific methods currently only accept a path parameter. As a result, users have to manually concatenate any query parameters to the path.

Modifications:

- Overload the get() method to accept both a path as well as a `QueryParams params` object.
- If the `QueryParams` object is either null or empty, do not concatenate to the path. If not empty, concatenate the original path, a `?` and the output of  `params.toQueryString()`. 

Result:

- Users can now access the API without manually concatenating query parameters to the path
- Closes #3915.

### CODE REVIEW

File: core/src/test/java/com/linecorp/armeria/client/DefaultWebClientTest.java [hunk 1]
```diff
@@ -89,16 +90,16 @@ void endpointRemapper() {
...
-                                          })
-                                          .build();
+                .endpointRemapper(endpoint -> {
+                    if ("group".equals(endpoint.host())) {
+                        return group;
+                    } else {
+                        return endpoint;
+                    }
+                })
+                .build();
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/client/DefaultWebClientTest.java:102]: Revert?

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java [hunk 2]
```diff
@@ -314,6 +315,19 @@ default HttpResponse get(String path) {
...
+     * Sends an HTTP GET request by appending the provided query params to the path.
+     */
+    @CheckReturnValue
+    default HttpResponse get(String path, QueryParams params) {
+        requireNonNull(params, "params");
+        if (params.isEmpty()) {
+            return this.get(path);
+        }
+
+        return execute(RequestHeaders.of(HttpMethod.GET, String.format("%s?%s", path, params.toQueryString())));
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/WebClient.java:328]: - If a path contains `?` already, we should concatenate a query string with `&`.
- The performance of `String.format()` rather slow than `+` or `StringBuilder`. Could you use `+` for merging two strings?
- It would be convenient to support `QueryParams` in other APIs such as `post()`, `patch()` and so on.

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java [hunk 3]
```diff
@@ -298,6 +299,25 @@ default WebClientRequestPreparation prepare() {
         return new WebClientRequestPreparation(this);
     }
 
+    /**
+     * Appends query params to the current path and returns the modified path.
+     */
+    default String addQueryParams(String path, QueryParams params) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/WebClient.java:305]: I'm not sure whether we need to expose this method as a public API.
How about hiding this by moving to a package-private class?
```
package com.linecorp.armeria.client;

final class WebClientUtil {

    static String addQueryParams(String path, QueryParams params) {
    ...
}
```

File: core/src/test/java/com/linecorp/armeria/client/DefaultWebClientTest.java [hunk 4]
```diff
@@ -64,16 +66,16 @@ void testWithoutRequestParamsUndefinedEndPoint() throws Exception {
...
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/client/DefaultWebClientTest.java:78]: Could you revert this change?
You might want to refer to this https://armeria.dev/community/developer-guide#setting-up-your-ide
to set up your IDE.

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java [hunk 5]
```diff
@@ -298,6 +299,25 @@ default WebClientRequestPreparation prepare() {
...
+        requireNonNull(params, "params");
+        if (params.isEmpty()) {
+            return path;
+        }
+
+        final StringBuilder appendedPath = new StringBuilder(path);
+        if (path.indexOf('?') == -1) {
+            appendedPath.append("?");
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/WebClient.java:313]:
```suggestion
            appendedPath.append('?');
```

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java [hunk 6]
```diff
@@ -298,6 +299,25 @@ default WebClientRequestPreparation prepare() {
...
+        } else {
+            appendedPath.append("&");
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/WebClient.java:315]:
```suggestion
            appendedPath.append('&');
```

**Review by minwoox — COMMENTED** [review]
Could you also add the same variants to `AbstractHttpRequestBuilder`?

**Comment by minwoox** [comment]
> Could you also add the same variants to AbstractHttpRequestBuilder?

Please forget about this comment because we have `queryParam` setters. 😄

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java (hunk 2 above)
Comment (nvidyala) [reply core/src/main/java/com/linecorp/armeria/client/WebClient.java]: I've pushed the mentioned changes, had a small doubt about code coverage: should I add tests for each of the overloaded methods that have QueryParams as a parameter? Writing those tests seems redundant to me as long as we're testing the core method appending the QueryParams string and integration with at least one method.

File: core/src/test/java/com/linecorp/armeria/client/DefaultWebClientTest.java (hunk 4 above)
Comment (nvidyala) [reply core/src/test/java/com/linecorp/armeria/client/DefaultWebClientTest.java:78]: Sorry about this happening again, I'll double-check if I imported the settings correctly.

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java (hunk 5 above)
Comment (nvidyala) [reply core/src/main/java/com/linecorp/armeria/client/WebClient.java:313]: Fixed.

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java (hunk 6 above)
Comment (nvidyala) [reply core/src/main/java/com/linecorp/armeria/client/WebClient.java:315]: Fixed.

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java (hunk 3 above)
Comment (nvidyala) [reply core/src/main/java/com/linecorp/armeria/client/WebClient.java:305]: Done.

File: core/src/test/java/com/linecorp/armeria/client/DefaultWebClientTest.java (hunk 1 above)
Comment (nvidyala) [reply core/src/test/java/com/linecorp/armeria/client/DefaultWebClientTest.java]: Fixed.

File: core/src/main/java/com/linecorp/armeria/client/WebClientUtil.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        final StringBuilder appendedPath = new StringBuilder(path);
+        if (path.indexOf('?') == -1) {
+            appendedPath.append('?');
+        } else {
+            appendedPath.append('&');
+        }
+        appendedPath.append(params.toQueryString());
+        return appendedPath.toString();
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/WebClientUtil.java:40]:
```suggestion
        return params.appendQueryString(appendedPath).toString();
```

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java (hunk 2 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/WebClient.java]: It might be tiresome to test all overloaded methods. Could you add integration tests for `GET` and `POST`, please?

Comment (nvidyala) [reply core/src/main/java/com/linecorp/armeria/client/WebClient.java]: I've added tests for the `GET` and `POST` methods of WebClient, although I'm not entirely sure if this is what you meant. Please let me know if I'm mistaken.

File: core/src/main/java/com/linecorp/armeria/client/WebClientUtil.java (hunk 7 above)
Comment (nvidyala) [reply core/src/main/java/com/linecorp/armeria/client/WebClientUtil.java]: Done.

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019 LINE Corporation
+ * Copyright 2021 LINE Corporation
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/WebClient.java:2]: nit: revert?
We don't change the year when we modify the code. 😉

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java [hunk 9]
```diff
@@ -330,6 +355,14 @@ default HttpResponse post(String path, HttpData content) {
         return execute(RequestHeaders.of(HttpMethod.POST, path), content);
     }
 
+    /**
+     * Sends an HTTP POST request with the specified content, accepts query parameters to append to the path.
+     */
+    @CheckReturnValue
+    default HttpResponse post(String path, HttpData content, QueryParams params) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/WebClient.java:362]: Probably better to do 
```suggestion
    default HttpResponse post(String path, QueryParams params, HttpData content) {
```
I feel that the `path` should be followed by the `params` because they are used in that way in the URL. 😄

**Review by minwoox — COMMENTED** [review]
Thanks! Left some minor nits. 😄

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java (hunk 8 above)
Comment (nvidyala) [reply core/src/main/java/com/linecorp/armeria/client/WebClient.java:2]: Done.

File: core/src/main/java/com/linecorp/armeria/client/WebClient.java (hunk 9 above)
Comment (nvidyala) [reply core/src/main/java/com/linecorp/armeria/client/WebClient.java:362]: Makes sense from a readability perspective, I've made the change for the overloaded methods of all verbs to include QueryParams after the path.

Comment (nvidyala) [reply core/src/main/java/com/linecorp/armeria/client/WebClient.java]: Please bear with me if there are any more. 😄

Also, I've made quite a few changes since the last commit for which a build was triggered, could you please approve a build run on these changes? Just wanted to ensure we're good coverage-wise.

**Comment by nvidyala** [comment]
> Thanks! Left some minor nits. 😄

Please bear with me if there are any more. 😄 

Also, quite a few changes have gone in since the last build. Could you please approve a build run? I'd like to verify that we're good coverage-wise.

File: core/src/main/java/com/linecorp/armeria/client/WebClientUtil.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+import com.linecorp.armeria.common.QueryParams;
+
+final class WebClientUtil {
+
+    /**
+     * Appends query params to the current path and returns the modified path.
+     */
+    static String addQueryParams(String path, QueryParams params) {
+        requireNonNull(params, "params");
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/WebClientUtil.java:28]: I guess we also need:
```suggestion
        requireNonNull(path, "path");
        requireNonNull(params, "params");
```

**Review by minwoox — APPROVED** [review]
Thanks a lot, @nvidyala!
Congratulations on your first contribution. 🎉 🎉  🎉

**Review by ikhoon — APPROVED** [review]
Thanks for your first contribution, @nvidyala!
I'm looking forward to your next step. 😄

File: core/src/main/java/com/linecorp/armeria/client/WebClientUtil.java (hunk 10 above)
Comment (nvidyala) [reply core/src/main/java/com/linecorp/armeria/client/WebClientUtil.java:29]: Done.

**Comment by minwoox** [comment]
Thanks and Congratulations on your first PR, @nvidyala! 🎉 🎉 🎉
